src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -35,6 +35,12 @@ public partial class Solution
         /// </summary>
         private AsyncLazy<Solution> CachedFrozenSolution { get; init; }
 
+        /// <summary>
+        /// Mapping of DocumentId to the frozen solution we produced for it the last time we were queried.  This
+        /// instance should be used as its own lock when reading or writing to it.
+        /// </summary>
+        private readonly Dictionary<DocumentId, AsyncLazy<Solution>> _documentIdToFrozenSolution = [];
```

Comment:
Note: 

1. i've been burned by ConcurrentDictionary in the past.  So i prefer just a normal dictionary with locking.  Could revisit this if we see any issues.
2. currently, it's just using normal locking.  we could consider cancellable locking with a semaphore slim if we see any issues.